packages/go_router/lib/src/route.dart

```diff
@@ -449,19 +460,296 @@ class ShellRoute extends RouteBase {
+///       },
+///       pageProvider:
+///           (BuildContext context, GoRouterState state, Widget statefulShell) {
+///         return NoTransitionPage<dynamic>(child: statefulShell);
+///       },
+///       routes: <GoRoute>[
+///         /// The first branch, i.e. root of tab 'A'
+///         GoRoute(
+///           parentNavigatorKey: _tabANavigatorKey,
+///           path: '/a',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'A', detailsPath: '/a/details'),
+///         ),
+///         /// The second branch, i.e. root of tab 'B'
+///         GoRoute(
+///           parentNavigatorKey: _tabBNavigatorKey,
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// To access the current state of this route, to for instance access the
+/// index of the current route branch - use the method
+/// [StatefulShellRoute.of]. For example:
+///
+/// ```
+/// final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example using StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
+  /// Constructs a [StatefulShellRoute] from a list of [ShellRouteBranch], each
+  /// representing a root in a stateful route branch.
+  ///
+  /// A separate [Navigator] will be created for each of the branches, using
+  /// the navigator key specified in [ShellRouteBranch].
+  StatefulShellRoute({
+    required this.branches,
+    required this.builder,
+    this.pageProvider,
+  })  : assert(branches.isNotEmpty),
+        assert(_uniqueNavigatorKeys(branches).length == branches.length,
+            'Navigator keys must be unique'),
+        super._(routes: _rootRoutes(branches)) {
+    for (int i = 0; i < routes.length; ++i) {
+      final RouteBase route = routes[i];
+      if (route is GoRoute) {
+        assert(route.parentNavigatorKey == null ||
+            route.parentNavigatorKey == branches[i].navigatorKey);
+      }
+    }
+  }
+
+  /// Constructs a [StatefulShellRoute] from a list of [GoRoute]s, each
+  /// representing a root in a stateful route branch.
+  ///
+  /// This constructor provides a shorthand form of creating a
+  /// StatefulShellRoute from a list of GoRoutes instead of
+  /// [ShellRouteBranch]s. Each GoRoute provides the navigator key
+  /// (via [GoRoute.parentNavigatorKey]) that will be used to create the
+  /// separate [Navigator]s for the routes.
+  StatefulShellRoute.rootRoutes({
+    required List<GoRoute> routes,
+    required ShellRouteBuilder builder,
+    ShellRoutePageBuilder? pageProvider,
+  }) : this(
+          branches: routes.map((GoRoute e) {
```

Comment:
Yes, perhaps, although you'll want to avoid duplication the assertion code. So perhaps we can instead use a private constructor (that also accepts routes), which the existing ones can call? Like this: 

```
  StatefulShellRoute._({
    required super.routes,
    required this.branches,
    required this.builder,
    this.pageProvider,
  })  : assert(branches.isNotEmpty),
        assert(_debugUniqueNavigatorKeys(branches).length == branches.length,
            'Navigator keys must be unique'),
        super._() {
    for (int i = 0; i < routes.length; ++i) {
      final RouteBase route = routes[i];
      if (route is GoRoute) {
        assert(route.parentNavigatorKey == null ||
            route.parentNavigatorKey == branches[i].navigatorKey);
      }
    }
  }
```